executor/src/lib.rs

```diff
@@ -752,72 +752,91 @@ where
+                        ensure!(
+                            !vm_output.write_set().is_empty(),
+                            "Transaction with empty write set should be discarded.",
                         );
+                        // Compute hash for the TransactionInfo object. We need the hash of the
+                        // transaction itself, the state root hash as well as the event root hash.
+                        let txn_info = TransactionInfo::new(
+                            txn.hash(),
+                            state_tree.root_hash(),
+                            event_tree.root_hash(),
+                            vm_output.gas_used(),
+                            status.major_status,
+                        );
+
+                        let real_txn_info_hash = txn_info.hash();
+                        txn_info_hashes.push(real_txn_info_hash);
+                        txn_info_hash = Some(real_txn_info_hash);
+                    }
+                    TransactionStatus::Discard(status) => {
+                        if !vm_output.write_set().is_empty() || !vm_output.events().is_empty() {
+                            crit!(
+                                "Discarded transaction has non-empty write set or events. \
+                                 Transaction: {:?}. Status: {}.",
+                                txn,
+                                status,
+                            );
+                        }
                     }
+                    TransactionStatus::Retry => (),
                 }
-            }
 
-            txn_data.push(TransactionData::new(
-                blobs,
-                vm_output.events().to_vec(),
-                vm_output.status().clone(),
-                Arc::clone(&state_tree),
-                Arc::new(event_tree),
-                vm_output.gas_used(),
-                num_accounts_created,
-                txn_info_hash,
-            ));
-            current_state_tree = state_tree;
-
-            // check for change in validator set
-            let validator_set_change_event_key = ValidatorSet::change_event_key();
-            for event in vm_output.events() {
-                if *event.key() == validator_set_change_event_key {
-                    next_validator_set = Some(ValidatorSet::from_bytes(event.event_data())?);
-                    break;
+                txn_data.push(TransactionData::new(
+                    blobs,
+                    vm_output.events().to_vec(),
+                    vm_output.status().clone(),
+                    Arc::clone(&state_tree),
+                    Arc::new(event_tree),
+                    vm_output.gas_used(),
+                    num_accounts_created,
+                    txn_info_hash,
+                ));
+                current_state_tree = state_tree;
+
+                // check for change in validator set
+                let validator_set_change_event_key = ValidatorSet::change_event_key();
```

Comment:
is this expensive? I doubt it since it should ideally be a constant function the compiler ought to make this unnecessary. If it is necessary, we probably ought to take it out of this for loop.